Finalizing the LupIO Devices

mkjost0 opened the pull request:
This is the cleaned up version of the current LupIO devices we support in gem5.  The only things not currently added in this pull request is the LupIO-IPI/SMP support, which we are currently finishing up, and the LupIO-SYS device, which I'm just trying to finish fixing some device tree issues with.  At every point in which a device is added to the LupVBoard, it should be able to run, given that the correct disk image and bootloader are provided.  The updated instructions on how to run this LupVBoard are in the README here (https://github.com/darchr/lupio-gem5/blob/configs/README.md), and the only file necessary from the lupio-gem5 repository to run the LupVBoard is the run_riscv.py file on the configs branch.  Once all these changes are finalized, these 6 devices should be ready to push upstream while we finish the last 2.  The only thing I wasn't sure about at this point was if we want to include the run_riscv.py file within gem5 itself, and if so, where.

lhinman19 committed:
dev: Added new LupIO-RTC device

This device supplies the computer system with the simulated time by
retrieving the current tick number and converting to seconds, then
returning that time in ISO 8601 format. It is implemented as a
BasicPioDevice and is read-only.

The following are the specifications regarding the LupIO-RTC:
https://gitlab.com/luplab/lupio/lupio-specs/-/blob/main/lupio-rtc.md

Change-Id: Ic99291e5c862e728dac90b15ca818d42f25cbbee

mkjost0 committed:
dev: Added new LupIO-RNG device

This device is a random number generator that uses the Mersenne
Twister in order to provide the system with a set of random
numbers.  It is implemented as a BasicPioDevice, and has both
read and write capabilities.

The following are the specifications regarding the LupIO-RNG:
https://gitlab.com/luplab/lupio/lupio-specs/-/blob/main/lupio-rng.md

Change-Id: Ia6aeb610ebe5589ed1f1548b823c5165236b03e6

lhinman19 committed:
dev: Added new LupIO-TTY device

This device is notfied when data is available from the terminal
(e.g. from keyboard input) in order to receive characters. It also
transmits characters to the terminal to be displayed.

The following are the specifications regarding the LupIO-TTY:
https://gitlab.com/luplab/lupio/lupio-specs/-/blob/main/lupio-tty.md

Change-Id: Icc8294984989cfa422d8ed227da39debfa49ab36

mkjost0 committed:
dev: Added new Lupio-BLK Device

This is a virtual block device that provides a disk-like interface
for second level storage.  It is implemented as a DMADevice, and
allows for the transfer of blocks from the block device to main
memory, and vice versa.

The following are the specifications regarding the LupIO-BLK:
https://gitlab.com/luplab/lupio/lupio-specs/-/blob/main/lupio-blk.md

Change-Id: Ifabc9b715fadb218e84952694d666b803e46e1f7

lhinman19 committed:
dev: Added new LupIO-TMR device

This device is a virtual timer that provides both a real-time
counter, as well as a configurable timer with periodic and
one-shot modes.  It uses Ticks to measure time, and is
implemented as a BasicPioDevice.

The following are the specifications regarding the LupIO-TMR:
https://gitlab.com/luplab/lupio/lupio-specs/-/blob/main/lupio-tmr.md

Change-Id: I6fd6f4926494a44d20e1e0289f502535e84d7a69

lhinman19 committed:
dev: Added new LupIO-PIC device

This device is a virtual programmable interrupt controller, and it
manages interrupt requests from up to 32 sources.  It is implemented
as a BasicPioDevice.

The following are the specifications regarding the LupIO-PIC:
https://gitlab.com/luplab/lupio/lupio-specs/-/blob/main/lupio-pic.md
Change-Id: I9ccdb607789f62cc89bdd7392d8e59c8e5c24797

mkjost0 committed:
dev,arch: Introduced new  LupV Platform

This is a platform with a RISC-V processor and the LupIO devices that will
allow users to decide which programmable interrupt controller to use in
their system.  It currently uses the PLIC device.

Change-Id: Ife6cf5c14845be725e66178693e9ba0ee5fda511

mkjost0 committed:
dev,arch: Introduced new LupV Board

This LupV Board was created in order to connect all of the LupIO
devices, and allow us to run a full RISC-V system with them.  As
the LupIO devices continue to be added, they will be integrated
into this board, and replace the current IO components.

The LupIO devices are a collection of processor
agnostic and easily implemented virtual devices. Details about the
specifications of all eight LupIO devices can be found here:
https://gitlab.com/luplab/lupio/lupio-specs

Information about how to build a RISCV full system with the LupIO-RTC
can be found here:
https://github.com/darchr/lupio-gem5/blob/lupio/README.md

Change-Id: I7d3186d3778d40b38027f245290432dbf4279dea

mkjost0 committed:
dev,arch: Added LupioRNG to LupVBoard

This adds the LupioRNG device to our LupVBoard, adding
random number generation capability to this RISC-V
based system.

Change-Id: I89c78c119a46cc7b056005d8abef16b1816ee4c4

mkjost0 committed:
dev,arch: Added LupioTTY to LupVBoard

This added the new LupioTTY device to the growing
LupVBoard, and replaced the Uart8250 device in this
system.

Change-Id: Ib9a09565e8522178fd9c9f900db9d3b87e8b48fb

mkjost0 committed:
dev,arch: Added LupioBLK Device to LupVBoard

This adds the LupioBLK device to the growing LupVBoard,
as well as removes the VirtIOMMIO Device from the board.

Change-Id: Ibb45ecff07c7ce5d5d7be8cd8e20e9f62b591b02

mkjost0 committed:
dev,arch: Modified LupV Platform + Board to use LupioPIC + TMR

This added both the LupioPIC and LupioTMR to the LupVBoard.  While
both the PLIC and CLINT are left in the board for the bootloader
to recognize, they aren't used within the system.  In addition, the
LupV Platform was changed in order to use the LupioPIC to handle
interrupts instead of the PLIC.

Change-Id: I57005903a7ec1136b42433ef5022ccb995abb9d6

mkjost0 committed:
dev: Modified LupioBLK and LupioTTY to use LupioPIC

This replaced the PLIC device within the LupioBLK and
LupioTTY with the LupioPIC, so that interrupts now go
through the LupioPIC, and the PLIC isn't used anymore.

Change-Id: I0eb5d5c5df9cb43cfb5e8e3a5bf4176f48320696

powerjg reviewed (commented):
A few small things.

Re: Joel's question... I like the idea of reviewers being able to see how each new device is used. That said, it's not terribly important to keep the history. I don't have strong opinions. I'd lean toward whatever Melissa thinks is best.

BobbyRBruce reviewed (commented):
This is a really fantastic piece of work. It mostly looks good to me, I only really have minor comments.

What i'd really like to see if an example script and some testing to ensure this code remains functional going forward. Here's  an example script I made for the X86Board (which also doubles up as a test): https://gem5-review.googlesource.com/c/public/gem5/+/50888. Maybe you could include something like this?

mkjost0 committed:
dev: Modify LupIO-TMR for SMP support

Added a new LupioTimer struct, as well as a timer event function for
SMP support.

Change-Id: Idbcc549dfa3c5f8d5342d7e2250337a7482a1ac0

mkjost0 committed:
dev: Modify LupIO-PIC for SMP support

Added mask and enable arrays to the LupIO-PIC in order to
handle SMP support.

Change-Id: Id120206f33374ce337e389ef5a10fd64e62bcab3

mkjost0 committed:
dev: Added new Lupio-IPI device

Added the LupIO inter-processor interrupt controller
device which will allow for us to use an SMP system
with the LupIO devices.

Change-Id: Iceab7446b36fb4d9b7605f3ba28665fca509d55d

mkjost0 committed:
dev,arch: Added LupIO-IPI to the LupV Board

This adds the LupioIPI device to our LupV Board, which
finalizes the changes needed for the LupIO devices to run
with SMP support.

Change-Id: I3e51dae6d8d589d3b38c57c33322e740c832a03d

mkjost0 commented on src/dev/lupio/lupio_tty.hh:
```diff
@@ -0,0 +1,101 @@
+    /**
+     * Inform the LupIO-TTY there is data available
+     */
+    void dataAvailable();
```
I don't think it does; I tried adding one in and it gave me some errors.

mkjost0 commented on src/dev/riscv/lupv.cc:
```diff
@@ -0,0 +1,86 @@
+    panic("LupV::pciToDma() has not been implemented.");
+}
+
+void
+LupV::serialize(CheckpointOut &cp) const
+{
+}
+
+void
+LupV::unserialize(CheckpointIn &cp)
+{
+}
```
I'm not sure, these functions were just from when we copied the format of the HiFive files from earlier on.  I tried commenting them out and it didn't seem to affect anything, so should I remove them, or leave them in still?

mkjost0 commented on src/python/gem5/components/boards/lupv_board.py:
```diff
@@ -0,0 +1,481 @@
+        cache_hierarchy: AbstractCacheHierarchy,
+    ) -> None:
+        super().__init__(clk_freq, processor, memory, cache_hierarchy)
+        if get_runtime_isa() != ISA.RISCV:
```
I've been trying to integrate this in, but I keep getting the exception below, so I've left it as is for now but keep looking into it.  
`Exception: [LupvBoard] The current ISA is 'RISCV'. Required: 'RISCV'`

BobbyRBruce replied:
Ok. This is not a big problem. It is an odd bug though.

BobbyRBruce replied:
I've figured out this bug. A fix is here: https://gem5-review.googlesource.com/c/public/gem5/+/53003

mkjost0 committed:
dev: Added new Lupio-SYS device

This adds the LupIO system controller, in which we provide a
way to halt or reboot the system.  It is implemented as a
BasicPIODevice.

Change-Id: I0031ac65d2aaca3460dcd4c59543a75230b0b52a

mkjost0 committed:
dev,arch: Added Lupio-SYS device to LupV Board

This adds the Lupio-SYS device to the LupV Board, which finalizes
all of the devices needed to complete the LupIO device
collection within gem5.

Change-Id: I5af34d8c2735d74f4240d3077ed162dd25f2bcb2

mkjost0 commented:
I added the LupIO-SYS device, so now the LupIO device collection should be complete within gem5! The only thing I was concerned about with the system controller is that it's supposed to be implemented as a write-only device, but running the `poweroff -f` command within the simulation is causing a read to be issued to the device.  I currently have it implemented so that the read and write functions are identical so it runs, but I wanted a second look at the code in case there was some reason I missed that's causing the device to be read from, instead of written to.

joel-porquet commented:
> I added the LupIO-SYS device, so now the LupIO device collection should be complete within gem5! The only thing I was concerned about with the system controller is that it's supposed to be implemented as a write-only device, but running the `poweroff -f` command within the simulation is causing a read to be issued to the device. I currently have it implemented so that the read and write functions are identical so it runs, but I wanted a second look at the code in case there was some reason I missed that's causing the device to be read from, instead of written to.

Thanks for bringing this up. I looked at the Linux kernel device driver for this device (it's a generic driver) and it does issue a read before writing a value. That's because, by default, it allows to only update certain bits to the system controller. Anyway, I've improved the specs to mention that reads are allowed, and they should return 0: https://gitlab.com/luplab/lupio/lupio-specs/-/blob/main/lupio-sys.md.

joel-porquet commented on src/python/gem5/components/boards/lupv_board.py:
```diff
@@ -143,9 +158,16 @@ def __init__(
-        self.pic.n_contexts = self.processor.get_num_cores() * 2
-        self.pic.n_src = max(pic_srcs) + 1
 
+        # Set the number of sources to the pic as 0 since we only have the
+        # PIC and CLINT connected to the board for our bbl to use upon 
+        # startup, and they will remain unused during the simulation
```
Well, technically the number of sources is 0 because there are no external devices actually connected to the PIC.

joel-porquet commented on src/dev/lupio/lupio_blk.hh:
```diff
@@ -32,16 +32,16 @@
 {
 
-class Plic;
+class LupioPIC;
```
Same question as before: do we actually need the blk and tty know about the PIC? I thought the interrupts were raised by calling the platform code directly, which then calls the PIC in a kind of generic way. Please correct me if I'm wrong.

joel-porquet commented on src/dev/lupio/lupio_sys.cc:
```diff
@@ -0,0 +1,117 @@
+                    addr);
+            break;
+    }
+    return 0;
```
So, for the read, and according to the fixed specs, you just want to return `0`. The actual exit/reboot should only be if the registers are being written to.

joel-porquet commented on src/dev/lupio/lupio_sys.cc:
```diff
@@ -0,0 +1,117 @@
+    switch (addr >> 2) {
+        case LUPIO_SYS_HALT:
+            DPRINTF(LupioSYS, "Trying to halt\n");
+            exitSimLoopNow("LUPIO_SYS_HALT called, exiting", 0, 0, false);
```
Is there no way to exit to the simulation loop with a specific exit code?

joel-porquet reviewed (commented):
So cool to see these 3 months worth of work condensed in ~20 commits! Looks really great. I made a few comments here and there, but the structure of the commits looks solid.

BobbyRBruce commented on src/python/gem5/components/boards/lupv_board.py:
```diff
@@ -0,0 +1,481 @@
+    FdtState,
+)
+
+class LupvBoard(SimpleBoard):
```
I think due to this change they'll be bugs if you rebase this on top of develop. I think the best course of action is to upload this to gerrit as it (well, once the other comments have been addressed), and I'll rebase this for you and ensure it works with the new changes.

mkjost0 commented on src/python/SConscript:
```diff
@@ -37,6 +37,7 @@ PySource('gem5.components.boards', 'gem5/components/boards/__init__.py')
 PySource('gem5.components.boards', 'gem5/components/boards/abstract_board.py')
 PySource('gem5.components.boards', 'gem5/components/boards/mem_mode.py')
 PySource('gem5.components.boards', 'gem5/components/boards/riscv_board.py')
+PySource('gem5.components.boards', 'gem5/components/boards/lupv_board.py')
```
For this, I updated the lupio-specs, but for the README, I was under the impression that we'd be putting this information somewhere else since this was supposed to be a temporary repo, but I'm not sure where it should go.  However, I'll go ahead and get rid of the second branch and make sure it's updated.

mkjost0 committed:
dev,arch: Moved LupV Board to an experimental folder

Change-Id: I9cd67177fe8f0cb34bebb4fd3f9d3af0a518c7b6

mkjost0 committed:
dev: Added LupV script to configs

This commit adds run_lupv.py, the script to run the LupVBoard, as well
as a README that details how to set up your system in order to run
the LupIO devices.

Change-Id: Ibea098f0409819b4c79063ab0be01518c7c6a61f

joel-porquet commented on configs/example/lupv/README.md:
```diff
@@ -0,0 +1,48 @@
+
+This README details how to build a RISCV full system that utilizes the
+lupIO devices. The LupIO devices were created by Professor Joël Porquet-Lupine,
+and more information about the device can be found here (https://luplab.cs.ucdavis.edu/assets/lupio/wcae21-porquet-lupio-paper.pdf).  The specs for each individual device can
```
This line looks a little long :)
